lib/layer.js

```diff
@@ -111,35 +109,31 @@ Layer.prototype.match = function match(path) {
   if (path != null) {
     // fast path non-ending match for / (any path matches)
     if (this.regexp.fast_slash) {
-      this.params = {}
-      this.path = ''
-      return true
+      return {
+        path: '',
+        params: {}
+      }
     }
 
     // fast path for * (everything matched in a param)
     if (this.regexp.fast_star) {
-      this.params = {'0': decode_param(path)}
-      this.path = path
-      return true
+      return {
+        path: path,
+        params: {'0': decode_param(path)}
+      }
```

Comment:
This code has been around in this state for a *very* long time. My worry is that this new way creates garbage to collect (the new object added) on a hot path, which might be the reason it was done this way in the first place. Maybe @dougwilson will remember because this predates me by quite a bit. Without more context I would like to see us do some sort of performance pass on this since I don't think the feature addition would be worth it if we impacted performance in a negative way.

Reply:
A big part was preventing bail out due to the same variable being different shapes (i.e. here we now have something that can be a bool or an object, and v8 will struggle to make a bytecode and may be falling back to slow interpreter mode). Ideally if we want to keep things in hot paths fast they should definitely keep the same shapes for the byte code generator.

I don't remember off hand all the details for why it was mutating properties, but it was very likely to reduce generating a bunch of little objects only to trash them immediately. That can be addressed if an object pool is used for them, however.

Reply:
I haven't run through a perf yet as I am away, but I suspect using `null` instead of `false` would fix any bail outs since those would both be objects

Reply:
Yeah that is a good point as well. I think ideally we have some real world benchmarks to base these decisions off of, but I think in theory both the de-opt from different shapes and increased GC from these little objects will be perf regressions. It would be nice to comment these with links to the benchmarks at some point so that the context is kept in the code for future contributors.

Reply:
So far what I have done is run the perf with the test suite, which the deopts from there usually pretty low hanging fruit.

Reply:
But ya, if it needs a shaped object can make a no match object at the top level and return that reference for the no match

Reply:
I made the change to use `null` instead of `false` for the no match scenario (I added a new commit to help reading the new changes, but I can squash them if you prefer).

Reply:
I am not sure I see how that addresses the worry here. The change you made does mean you could return some of those boolean checks back to their original form I think, which is overall good to reduce the change for change's sake, but I think this comment thread here is about the method returning a boolean and not producing unnecessary garbage in the process.

I am not suggesting removing the last changes, but I think ideally we find a way to not create these objects yet still clean up the state after the layers are finished being handled.